### ISSUE TITLE
A styling span (bold/italic) can cover more than one paragraph, causing an unbalanced stack

### ISSUE THREAD

**Issue by Azeirah** [ricklupton/rmscene#22]
You can see the problem on... 

- page 1 on the attached RM document  `d133d282-93aa-4dfd-a7fb-00eba9a0f8d3.rm` 
- as well as on page 5 `1c08e4c9-f45e-4e73-8c5f-bfeb3bac42ad.rm'` (this one is clearer, I made this page to intentionally showcase the problem)

The code in `text.py` assumes that the bold/italic stack should be balanced per paragraph, but this is not necessarily the case. The stack is balanced over the entire text document. This is because italic or bold can span multiple paragraphs. As you can see on the attached images, I have three paragraphs of text in a row which all have italic styling applied.

The following two images show the italicized text and the selection I made to create the italic span respectively.

![WhatsApp Image 2023-09-23 at 13 06 09](https://github.com/ricklupton/rmscene/assets/4040870/23776752-0ac3-4477-9365-64c54e0798d6)

![WhatsApp Image 2023-09-23 at 13 06 10](https://github.com/ricklupton/rmscene/assets/4040870/79c6532f-1adc-49cc-a270-51a44616b229)

## The per-character log

This log is for page 5.

I added print statements to `TextDocument.from_scene_item`, printing out each character per iteration.

(Note I replaced the deleted character with "キ" to make them stand out during my debugging, they have no special meaning other than being a deleted character)

As you can see in the log, the italic span opens in the first paragraph at `CrdtId(1, 151)` and ends at the third paragraph at `CrdtId(1, 152)`

```
CrdtId(1, 15) "キ"
CrdtId(1, 16) "キ"
CrdtId(1, 17) "キ"
CrdtId(1, 18) "キ"
CrdtId(1, 19) "キ"
CrdtId(1, 20) "キ"
CrdtId(1, 21) "キ"
CrdtId(1, 22) "キ"
CrdtId(1, 23) "キ"
CrdtId(1, 24) "キ"
CrdtId(1, 25) "キ"
CrdtId(1, 26) "キ"
CrdtId(1, 27) "キ"
CrdtId(1, 28) "キ"
CrdtId(1, 29) "キ"
CrdtId(1, 30) "T"
CrdtId(1, 31) "h"
CrdtId(1, 32) "r"
CrdtId(1, 33) "e"
CrdtId(1, 34) "e"
CrdtId(1, 35) " "
CrdtId(1, 36) "キ"
CrdtId(1, 37) "キ"
CrdtId(1, 38) "キ"
CrdtId(1, 39) "キ"
CrdtId(1, 40) "キ"
CrdtId(1, 41) "キ"
CrdtId(1, 42) "キ"
CrdtId(1, 151) 3
CrdtId(1, 43) "p"
CrdtId(1, 44) "a"
CrdtId(1, 45) "r"
CrdtId(1, 46) "a"
CrdtId(1, 47) "g"
CrdtId(1, 48) "r"
CrdtId(1, 49) "a"
CrdtId(1, 50) "p"
CrdtId(1, 51) "h"
CrdtId(1, 52) "s"
CrdtId(1, 53) " "
CrdtId(1, 54) "w"
CrdtId(1, 55) "i"
CrdtId(1, 56) "t"
CrdtId(1, 57) "h"
CrdtId(1, 58) " "
CrdtId(1, 59) "i"
CrdtId(1, 60) "t"
CrdtId(1, 61) "a"
CrdtId(1, 62) "l"
CrdtId(1, 63) "i"
CrdtId(1, 64) "キ"
CrdtId(1, 65) "キ"
CrdtId(1, 66) "キ"
CrdtId(1, 67) "キ"
CrdtId(1, 68) "キ"
CrdtId(1, 69) "キ"
CrdtId(1, 70) "キ"
CrdtId(1, 71) "c"
CrdtId(1, 72) " "
CrdtId(1, 73) "i"
CrdtId(1, 74) "n"
CrdtId(1, 75) " "
CrdtId(1, 76) "o"
CrdtId(1, 77) "n"
CrdtId(1, 153) "e"
CrdtId(1, 78) " "
CrdtId(1, 79) "s"
CrdtId(1, 80) "e"
CrdtId(1, 81) "l"
CrdtId(1, 82) "e"
CrdtId(1, 83) "キ"
CrdtId(1, 84) "キ"
CrdtId(1, 85) "c"
CrdtId(1, 86) "t"
CrdtId(1, 87) "i"
CrdtId(1, 88) "o"
CrdtId(1, 89) "n"
CrdtId(1, 90) "
"
End of paragraph
Breaking
Unbalanced stack! [(None, [CrdtStr(s='キキキキキキキキキキキキキキキThree キキキキキキキ', i=[CrdtId(1, 15), CrdtId(1, 16), CrdtId(1, 17), CrdtId(1, 18), CrdtId(1, 19), CrdtId(1, 20), CrdtId(1, 21), CrdtId(1, 22), CrdtId(1, 23), CrdtId(1, 24), CrdtId(1, 25), CrdtId(1, 26), CrdtId(1, 27), CrdtId(1, 28Id(1, 30), CrdtId(1, 31), CrdtId(1, 32), CrdtId(1, 33), CrdtId(1, 34), CrdtId(1, 35), CrdtId(1, 36), CrdtId(1, 37), CrdtId(1, 38), CrdtId(1, 39), CrdtId(1, 40), CrdtId(1, 41), CrdtId(1, 42)])]), (<class 'rmscene.text.ItalicSpan'>, [CrdtStr(s='paragraphs with italiキキキキキキキc in one seleキキction'Id(1, 43), CrdtId(1, 44), CrdtId(1, 45), CrdtId(1, 46), CrdtId(1, 47), CrdtId(1, 48), CrdtId(1, 49), CrdtId(1, 50), CrdtId(1, 51), CrdtId(1, 52), CrdtId(1, 53), CrdtId(1, 54), CrdtId(1, 55), CrdtId(1, 56), CrdtId(1, 57), CrdtId(1, 58), CrdtId(1, 59), CrdtId(1, 60), CrdtId(1, 61), CrdtId(1, 62), CrdtId(1, 63), CrdtId(1, 64), CrdtId(1, 65), CrdtId(1, 66), CrdtId(1, 67), CrdtId(1, 68), CrdtId(1, 69), CrdtId(1, 70), CrdtId(1, 71), CrdtId(1, 72), CrdtId(1, 73), CrdtId(1, 74), CrdtId(1, 75), CrdtId(1, 76), CrdtId(1, 77), CrdtId(1, 153), CrdtId(1, 78), CrdtId(1, 79), CrdtId(1, 80), CrdtId(1, 81), CrdtId(1, 82), CrdtId(1, 83), CrdtId(1, 84), CrdtId(1, 85), CrdtId(1, 86), CrdtId(1, 87), CrdtId(1, 88), CrdtId(1, 89)])])]
-------------
New paragraph
-------------
CrdtId(1, 91) "T"
CrdtId(1, 92) "h"
CrdtId(1, 93) "i"
CrdtId(1, 94) "s"
CrdtId(1, 95) " "
CrdtId(1, 96) "i"
CrdtId(1, 97) "s"
CrdtId(1, 98) " "
CrdtId(1, 99) "t"
CrdtId(1, 100) "h"
CrdtId(1, 101) "e"
CrdtId(1, 102) " "
CrdtId(1, 103) "m"
CrdtId(1, 104) "i"
CrdtId(1, 105) "d"
CrdtId(1, 106) "d"
CrdtId(1, 107) "l"
CrdtId(1, 108) "e"
CrdtId(1, 109) " "
CrdtId(1, 110) "p"
CrdtId(1, 111) "a"
CrdtId(1, 112) "r"
CrdtId(1, 113) "a"
CrdtId(1, 114) "g"
CrdtId(1, 115) "r"
CrdtId(1, 116) "a"
CrdtId(1, 117) "p"
CrdtId(1, 118) "h"
CrdtId(1, 119) "
"
End of paragraph
Breaking
-------------
New paragraph
-------------
CrdtId(1, 120) "T"
CrdtId(1, 121) "H"
CrdtId(1, 122) "i"
CrdtId(1, 123) "s"
CrdtId(1, 124) " "
CrdtId(1, 125) "i"
CrdtId(1, 126) "s"
CrdtId(1, 127) " "
CrdtId(1, 128) "t"
CrdtId(1, 129) "h"
CrdtId(1, 130) "e"
CrdtId(1, 131) " "
CrdtId(1, 132) "e"
CrdtId(1, 133) "n"
CrdtId(1, 134) "d"
CrdtId(1, 135) " "
CrdtId(1, 136) "キ"
CrdtId(1, 137) "キ"
CrdtId(1, 138) "キ"
CrdtId(1, 139) "p"
CrdtId(1, 140) "a"
CrdtId(1, 141) "r"
CrdtId(1, 142) "a"
CrdtId(1, 143) "キ"
CrdtId(1, 144) "キ"
CrdtId(1, 145) "キ"
CrdtId(1, 146) "g"
CrdtId(1, 147) "r"
CrdtId(1, 148) "a"
CrdtId(1, 152) 4
Unexpected end of span at CrdtId(1, 152): got <class 'rmscene.text.ItalicSpan'>, expected None
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/lb/PhpstormProjects/rm-notesync-laravel/binaries/remarks/.venv/lib/python3.10/site-packages/rmscene/text.py", line 179, in from_scene_item
    contents = parse_paragraph_contents()
  File "/home/lb/PhpstormProjects/rm-notesync-laravel/binaries/remarks/.venv/lib/python3.10/site-packages/rmscene/text.py", line 154, in parse_paragraph_contents
    stack[-1][1].append(span_type(nested))
IndexError: list index out of range
```

[Text formatting text notebook](https://github.com/ricklupton/rmscene/files/12706267/f3879ca091badaabf83cdc304d0dcfaa1de1d854.zip)

**Comment by Azeirah** [comment]
In my code where I was experimenting with printing text to PDF in Remarks, I tried using a state machine rather than a hierarchy. This approach gets the styling correct.

In pseudocode

```
bold = False
italic = False
for char in chars:
    if char == 1:
        bold = True
    elif char == 2:
        bold = False
    elif char == 3:
        italic = True
    elif char == 4:
        italic = False
    # ... print char
```

If I'm correct, ReMarkable's text-format assumes the following:

- Each newline introduces a new paragraph
- 1 and 3 open bold and italic styling respectively
- 2 and 4 close bold and italic styling respectively

Expressed in hierarchical HTML, my text document looks like

```html
<h1>Three <i>paragraphs with italic in one selection</h1>
<p>This is the middle paragraph</p>
<p>This is the end paragraph</p>
</i>
```

While it's "possible" to mark up HTML in this way, it's unintuitive as well as incorrect. I'm not convinced the hierarchical data-structure is the correct data-structure for this text.

My suggestion would be to redesign the text document data structure as a list of paragraphs with special tokens marking the start and end of the stylized spans meant for consumption with a state machine printer/formatter.

IE:

```
bold = False
italic = False
for paragraph in doc.paragraphs:
    for char in chars:
        if char == 1:
            bold = True
        elif char == 2:
            bold = False
        elif char == 3:
            italic = True
        elif char == 4:
            italic = False
        # ... print char with appropriate styling
```

**Comment by ricklupton** [comment]
Thanks for the clear example!

I guess the other option would be to produce the hierarchical output equivalent to:

```html
<h1>Three <i>paragraphs with italic in one selection</i></h1>
<p><i>This is the middle paragraph</i></p>
<p><i>This is the end paragraph</i></p>
```

which might be easier for further processing? It would clearly make it easier to convert the output to HTML and similar output formats that expect the styles to fall within paragraphs. I'm not sure which is better for your converting to PDF?

(e.g. by using a similar approach internally to what you suggest, but closing and re-opening italic spans for each new paragraph as needed)

**Comment by Azeirah** [comment]
That's definitely a good option as well! For what it's worth I'm also looking to convert to markdown, and I'm sure there will be other use cases for other people as well. A valid hierarchy equivalent to HTML sounds perfect to me to be honest!

**Comment by ricklupton** [comment]
Fixed in e0ea16ba950f360638e139dfe82f1959903227b9

In the end I decided the hierarchical structure was too complicated, but it was useful if the output of rmscene had already interpreted the mysterious formatting codes. The updated version returns strings with text properties like `{"font-weight": "bold", "font-style": "normal"}` applied to the relevant sections. Does that work ok for your case @Azeirah ?

**Comment by Azeirah** [comment]
Absolutely! I'll try it out this weekend and get back to you.
